File: proposed-projects/round-trip-experiment/reverse/20260909T002206Z_273cf6b9/20260910T042804Z_e72f6c04/ported_implementation/gbnf/rules_builder/rules_builder_test.py
"""Cases for the rules builder.

The table is ~1400 lines of `InternalRuleDef*` literals, so it lives beside this file as
`rules_builder_cases.json` and is rebuilt into rule defs here.
"""

import json
from pathlib import Path

import pytest

from .rules_builder import RulesBuilder
from .rules_builder_types import (
    InternalRuleDefAlt,
    InternalRuleDefChar,
    InternalRuleDefCharAlt,
    InternalRuleDefCharNot,
    InternalRuleDefCharRngUpper,
    InternalRuleDefEnd,
    InternalRuleDefReference,
)

CASES = json.loads((Path(__file__).parent / "rules_builder_cases.json").read_text())

BUILDERS = {
    "char": InternalRuleDefChar,
    "char_alt": InternalRuleDefCharAlt,
    "char_not": InternalRuleDefCharNot,
    "char_rng_upper": InternalRuleDefCharRngUpper,
    "ref": InternalRuleDefReference,
    "alt": InternalRuleDefAlt,
    "end": InternalRuleDefEnd,
}


def build_rule_def(fixture):
    kind = fixture["kind"]
    if kind not in BUILDERS:
        raise Exception(f"Unknown rule def kind: {kind}")
    if "value" in fixture:
        return BUILDERS[kind](fixture["value"])
    return BUILDERS[kind]()


@pytest.mark.parametrize(
    ("key", "grammar", "expected"),
    [(key, grammar, expected) for key, grammar, expected in CASES],
)
def test_it_parses_a_grammar(key, grammar, expected):
    symbol_ids_expected, rules_expected = expected
    parsed_grammar = RulesBuilder(grammar.replace("\\n", "\n"))
    assert parsed_grammar.rules == [
        [build_rule_def(rule_def) for rule_def in rule] for rule in rules_expected
    ]
    assert list(parsed_grammar.symbol_ids.entries()) == [
        (name, id) for name, id in symbol_ids_expected
    ]

File: packages/gbnf-experiment/data/20260909T002206Z_273cf6b9/ported_implementation/tests/unit/rules-builder/rules-builder.test.ts
import { describe, expect, test } from 'vitest';
import { RulesBuilder } from '../../../src/rules-builder/rules-builder.ts';
import {
  InternalRuleDefAlt,
  InternalRuleDefChar,
  InternalRuleDefCharAlt,
  InternalRuleDefCharNot,
  InternalRuleDefCharRngUpper,
  InternalRuleDefEnd,
  InternalRuleDefReference,
  type InternalRuleDef,
} from '../../../src/rules-builder/rules-builder-types.ts';
import cases from '../../fixtures/rules-builder.json' with { type: 'json' };

interface RuleDefFixture {
  kind: string;
  value?: number | number[];
}

type Case = [string, string, [[string, number][], RuleDefFixture[][]]];

const buildRuleDef = ({ kind, value }: RuleDefFixture): InternalRuleDef => {
  switch (kind) {
    case 'char':
      return new InternalRuleDefChar(value as number[]);
    case 'char_alt':
      return new InternalRuleDefCharAlt(value as number);
    case 'char_not':
      return new InternalRuleDefCharNot(value as number[]);
    case 'char_rng_upper':
      return new InternalRuleDefCharRngUpper(value as number);
    case 'ref':
      return new InternalRuleDefReference(value as number);
    case 'alt':
      return new InternalRuleDefAlt();
    case 'end':
      return new InternalRuleDefEnd();
    default:
      throw new Error(`Unknown rule def kind: ${kind}`);
  }
};

describe('grammar parser', () => {
  test.each((cases as Case[]).map(([key, grammar, expected]) => [key, grammar, expected] as const))(
    '%s',
    (_key, grammar, expected) => {
      const [symbolIdsExpected, rulesExpected] = expected;
      const parsedGrammar = new RulesBuilder(grammar.replaceAll('\\n', '\n'));
      expect(parsedGrammar.rules).toStrictEqual(rulesExpected.map(rule => rule.map(buildRuleDef)));
      expect([...parsedGrammar.symbolIds.entries()]).toStrictEqual(
        symbolIdsExpected.map(([name, id]) => [name, id]),
      );
    },
  );
});
